Build guide: validation (#520)
holochain/docs-pages: src/pages/build/validate-callback.md

:::

## Define a `validate` callback

A `validate` callback takes an input argument of [`Op`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/op/enum.Op.html) and returns a [`ValidateCallbackResult`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/validate/enum.ValidateCallbackResult.html) wrapped in an `ExternResult`.

Validating an op can have one of three outcomes: {#validation-outcomes}

1. `ValidateCallbackResult::Valid` means that validation succeeded.
2. `ValidateCallbackResult::Invalid(String)` carries information about the validation failure for debugging.
3. <code>ValidateCallbackResult::UnresolvedDependencies([UnresolvedDependencies](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/validate/enum.UnresolvedDependencies.html))</code> means that validation couldn't finish because the required dependencies couldn't be fetched from the DHT.

> Outcomes 1 and 2 are definitive and the op won't be validated again. Outcome 3, which happens automatically when a call any of the [`must_get_*` functions](https://docs.rs/hdi/latest/hdi/entry/index.html#functions) yields no data, asks Holochain to try running the op through validation later in the hope that it can retrieve the required dependencies. (It'll give up retrying after some time.)

Reviewer: It does give up to save resources but if you restart Holochain it will resume trying. Abandon is not implemented as a validation outcome.

Author: oh! didn't know about either of those things.

Author: https://github.com/holochain/docs-pages/pull/520/commits/15dae3589a940e21125190542cdc9f1040fa2ee7#diff-6292d0d84462fb21a3681bc06fadd4f6cc525f5bc93015e51e25adf0e0c2b622R19

The simplest `validate` callback approves everything:

```rust
use hdi::prelude::*;

#[hdk_extern]
pub fn validate(_: Op) -> ExternResult<ValidateCallbackResult> {
    Ok(ValidateCallbackResult::Valid)
}
```

We show it here, not because it's useful, but so you can see the function's signature. Even less useful is one that does the opposite:

```rust
use hdi::prelude::*;

#[hdk_extern]
pub fn validate(_: Op) -> ExternResult<ValidateCallbackResult> {
    Ok(ValidateCallbackResult::Invalid("I reject everything".into()))
}
```

There is something worth noting here though. We return an `Ok` even though validation failed. That's because **`Err` should be reserved for true failures** such as the ones returned by host functions.

## Create boilerplate code with the scaffolding tool

> Even after flattening, there are a lot of variants of an op, which makes for a lot of match arms in your `validate` callback. And a lot of the arms do similar but slightly different things. It's a great idea to let the scaffolding tool generate the callback for you. We won't paste a sample here --- you can scaffold an integrity zome and a few entry and link types and see what it looks like.

Reviewer: Was flattening introduced before here?

Author: whoops, casualty of a reorganisation

Author: https://github.com/holochain/docs-pages/pull/520/commits/15dae3589a940e21125190542cdc9f1040fa2ee7#diff-6292d0d84462fb21a3681bc06fadd4f6cc525f5bc93015e51e25adf0e0c2b622R47-R49 , also includes some removal of redundant prose

Here are some useful examples that show you how to use the stub functions, imagining you scaffolded the `Director` and `Movie` entry types from the [Entries page](/build/entries/#define-an-entry-type) along with a collection for all `Director` entries.

### Stub functions for entries

These functions can be found in the file `dnas/<dna>/zomes/integrity/<zome>/src/<entry>.rs`.

#### `validate_create_<entry>`

In this function you can write rules for the contents of your entries and their dependencies. You can also write rules for the actions that write them, which means you can create rules for write privileges. (Note that it's called for both kinds of [`EntryCreationAction`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/op/enum.EntryCreationAction.html) --- `Create` and `Update`.)

This example checks that a movie is within [sensible bounds](https://en.wikipedia.org/wiki/Roundhay_Garden_Scene).

```rust
use hdi::prelude::*;

pub fn validate_create_movie(
    _action: EntryCreationAction,
    movie: Movie,
) -> ExternResult<ValidateCallbackResult> {
    // Note that converting a `&str` to a `Timestamp` requires you to list

>     // `kitsune_p2p_timestamp` as a dependency in your `Cargo.toml`.

Reviewer 2: I think we'll remove or at least move `Timestamp` to a different place, so I suggest to use a std `Instant` in the example instead.

Reviewer: This has now moved to `holochain_timestamp`, I think it's probably okay to reference that now?

Author: @ThetaSinner has it moved there for the 0.4 series, or just 0.5?

Reviewer: Good point, only on 0.5

Reviewer: Although for this purpose, it doesn't really matter, you'd just have to compare the raw times with `.as_micros` because you wouldn't be able to compare a K1 timestamp with a Holochain timestamp. Maybe best left alone for now

Author: Oh wait -- there's a Holochain timestamp type? I thought there was only a K1 timestamp that was re-exported by `holochain_integrity_types`

Reviewer: Yes but K1 is being deleted, so the type has been moved. We should never have been using a Kitsune type in hApp code.

>     if movie.release_date < "1888-10-14".try_into().unwrap() {

Reviewer: Rather than having to parse a date for every movie that gets created, I'd figure out the UNIX timestamp for that data and put it in a constant. Then it's just `Timestamp::from_micros`

Author: So far in the build guide we've been turning strings into kitsune timestamps for legibility's sake; if we decide to change that, I'd like to do it in a separate PR.

Reviewer: Something like 

```rust
/// The timestamp for the date 1888-10-14, at 00:00
const EARLIEST_MOVIE_TIMESTAMP: Timestamp = Timestamp(1209483914);
```

Seems perfectly legible to me. I wouldn't reject it on a PR into Holochain.

Author: https://github.com/holochain/docs-pages/pull/520/commits/afe9fe11bf816defd19aaeb941a4868798646a61

        return Ok(ValidateCallbackResult::Invalid("The movie's release date is earlier than the oldest known film.".into()));
    }
    Ok(ValidateCallbackResult::Valid)
}
```

This example checks that a director entry for a movie exists.

```rust
use hdi::prelude::*;

pub fn validate_create_movie(
    _action: EntryCreationAction,
    movie: Movie,
) -> ExternResult<ValidateCallbackResult> {
    // Just call the function, and Holochain will handle the
    // `UnresolvedDependencies` outcome for you.
    let director_entry =  must_get_entry(movie.director_hash)?;
    // Try to turn it into an entry of the right type.

>     let _director = crate::Director::try_from(director_entry)?;

Reviewer: Has this been tested? I always struggle to keep how validation works in my head but I think this is expressing the wrong thing. This will be treated as a code error and *not* as failed validation. I think you need to explicitly handle this error and return an app validation rejected outcome if this doesn't point to a Director type. 

You could check what happens here if you write a coordinator function that tries to reference another movie in the `director_hash` for example

Author: hm, yeah, that should be a validation failure, not an error. Pretty sure it's gonna do what you think it's gonna do.

Reviewer: The `let = ... else` syntax will only do something useful if the assignment is doing pattern matching. So doesn't that need to be `let Ok(_) = .... else`? I'd expect the compiler to say the else is unreachable but maybe there's something I don't know there.

Linking diffs is helping but when there's that much changed in one commit it's less helpful. I'm assuming that commit went up before we discussed posting diff links but just FYI :)

Author: already caught that and fixed in https://github.com/holochain/docs-pages/pull/520/commits/11aa2217a77ee4da99a446763d4299b5bade0915#diff-6292d0d84462fb21a3681bc06fadd4f6cc525f5bc93015e51e25adf0e0c2b622L92-R95 .

I don't think I caught the request to link to individual commits rather than diffs, but I've been independently starting to move towards tiny commits that only change one thing, cuz it seems like it fits better with this sort of pattern.

    Ok(ValidateCallbackResult::Valid)
}
```

#### `validate_update_<entry>`

> This callback gets the original entry and its creation action along with the update. You can use it to compare differences between the two entries, or you can use it to enforce write permissions, such as only allowing the original author to update an entry:

Reviewer: Definitely worth noting that people should consider what this logic means. For a profile picture, it makes a lot of sense to restrict updates to the original author. For movies, I could claim the name of the movie in the app because I heard about it being filmed before it was released. Just a quick note here about the app validation rules being part of the business logic to an extent.

OR

In this example, you'd probably want to check that the create rules aren't violated by the update, and allow other people to correct wrong movie details as long as they follow the rules.

Author: The scaffolded `validate` callback calls `validate_create_movie` for a `StoreEntry` op and both `validate_create_movie` and `validate_update_movie` for a `StoreRecord` op, so I think it'd be redundant to have the example check that the create rules aren't violated. I'd like to show something that's unique to updates... you're right that this is an anti-pattern if you're worried about squatters though. Maybe I'll show an example using `MovieLoan` from the 'Identifiers' build guide instead.

    // entry creation action of a type called `invite`.
    Ok(ValidateCallbackResult::Valid)
}

#[hdk_extern]
pub fn genesis_self_check(data: GenesisSelfCheckData) -> ExternResult<ValidateCallbackResult> {
    validate_invite_code_format(data.membrane_proof.map(|b| b.bytes().to_owned()))
}

fn decode_invite_code(invite_code: Vec<u8>) -> ExternResult<ActionHash> {
    // Try to convert the invite code from bytes to an action hash.
    let invite_code = BASE64_STANDARD.decode(invite_code)
        .map_err(|e| wasm_error!(e.to_string()))?;
    let invite_code = std::str::from_utf8(&invite_code)
        .map_err(|e| wasm_error!(e.to_string()))?;
    let invite_code_action_hash = ActionHashB64::from_b64_str(invite_code)
        .map_err(|e| wasm_error!(e.to_string()))?;
    Ok(invite_code_action_hash.into())
}

fn validate_invite_code_format(invite_code: Option<Vec<u8>>) -> ExternResult<ValidateCallbackResult> {
    match invite_code {
        Some(invite_code) => {
            decode_invite_code(invite_code)?;
            Ok(ValidateCallbackResult::Valid)
        }
        None => Ok(ValidateCallbackResult::Invalid("Please supply an invite code.".into()))
    }
}
```

> <!-- TODO: Write about inductive validation when it's not broken https://github.com/holochain/holochain/issues/4669 -->

Reviewer: I don't follow. Validation is inductive when `must_get_valid_x` is used is it not?

Warrants are unstable, packed with bugs and disabled by default. I wouldn't let them stop you writing about the logic of depending on existing valid data.

Author: Okay sure, so it's not something you're worried about overpromising on? For me the concern is not whether warrants are buggy; it's my concern that, even once warrant bugs are fixed, `must_get_valid_record` is inherently non-deterministic -- we're choosing to relinquish our right to validate the dep and trust the word of a potentially Byzantine-faulty peer.

Reviewer: We never trust a single peer on a warrant, that's not the right way to think about it. If somebody who is an authority for some data sees a warrant then they perform the same validation that the author of the warrant did. If that validation also results in them wanting to issue a warrant then they are supposed to block the author of the original data. So what should happen is that all authorities for an invalid record end up with a warrant and block the author. This means that anybody who isn't an authority for that data can't get the record from any authority. There should be no case where we just trust a single warrant issuer. The point of treating it as unresolved dependencies is to see if we can retrieve that dependency from somebody else if one agent falsly claims a warrant. Whether it's correctly implemented is another question but we spent a lot of time puzzling out these cases.

I want to forget completely about warrants here. They're disabled until further notice. The logic of "an update is only valid only against a valid create" is entirely implementable today on Holochain by using `must_get_valid_record`.  I don't see why you'd need to re-check the create when validating an update. Using `must_get_valid_record` is the right way to do that.

Reviewer: There's nothing much we can do about that. I think it's acceptable for data that is otherwise valid but happens to depend on invalid data to be marked valid. Other agents might not be able to validate it to check your work but that doesn't seem too serious. It's ultimately business logic for the application to decide how/whether to prune out data that ended up relying on invalid data I think. Whether that's practical is another issue. This is very much on the fringes and outside what most apps need to worry too much about I think.

Author: added a note about inductive validation, and resolved todos that reminded me to link to it: https://github.com/holochain/docs-pages/pull/520/commits/58e879857a5df243bf7e18d233c9ac944120d321

Fortunately, the scaffolding tool generates `validate` and [`genesis_self_check`](/build/genesis-self-check-callback/) callbacks that call out to stub functions that you can fill in with your own validation logic.

Here are some useful examples that show you how to use the stub functions, imagining you've scaffolded the `Director` and `Movie` entry types from the [Entries](/build/entries/#define-an-entry-type) and the `MovieLoan` entry type from the [Identifiers](/build/identifiers/) page, along with a [global collection](/build/links-paths-and-anchors/#scaffold-a-simple-collection-anchor) for all `Director` entries.

### Stub functions for entries

These functions can be found in the file `dnas/<dna>/zomes/integrity/<zome>/src/<entry>.rs`.

#### `validate_create_<entry>`

In this function you can write rules for the contents of your entries and their dependencies. You can also write rules for the actions that write them, which means you can create rules for write privileges. (Note that it's called for both kinds of [`EntryCreationAction`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/op/enum.EntryCreationAction.html) --- `Create` and `Update`.)

> This example checks that a movie is within [sensible bounds](https://en.wikipedia.org/wiki/Roundhay_Garden_Scene).

Reviewer 3: ```suggestion
This example checks that a movie's release date is within [sensible bounds](https://en.wikipedia.org/wiki/Roundhay_Garden_Scene).
```

Author: ```suggestion
This example checks that a movie's release date is within [sensible bounds](https://en.wikipedia.org/wiki/Roundhay_Garden_Scene).
```

Can we meet halfway?

Reviewer 3: I feel like linking to a wikipedia page about the first movie ever made just adds noise.

```rust
use hdi::prelude::*;

// 14 October 1888
const EARLIEST_MOVIE_TIMESTAMP: Timestamp = Timestamp(-2562883200_000_000);

pub fn validate_create_movie(
    _action: EntryCreationAction,
    movie: Movie,
) -> ExternResult<ValidateCallbackResult> {
    if movie.release_date < EARLIEST_MOVIE_TIMESTAMP {
        return Ok(ValidateCallbackResult::Invalid("The movie's release date is earlier than the oldest known film.".into()));
    }
    Ok(ValidateCallbackResult::Valid)
}
```

This example checks that a director entry referenced by a movie exists.

```rust
use hdi::prelude::*;

pub fn validate_create_movie(
    _action: EntryCreationAction,
    movie: Movie,
) -> ExternResult<ValidateCallbackResult> {
    // Just call the function, and Holochain will handle the
    // `UnresolvedDependencies` outcome for you.

>     let director_entry =  must_get_entry(movie.director_hash)?;

Reviewer 3: ```suggestion
    let director_entry = must_get_entry(movie.director_hash)?;
```

    ))
}
```

And this example once again only allows people to delete movie loans they created:

```rust
use hdi::prelude::*;

pub fn validate_delete_movie_loan(
    action: Delete,
    original_action: EntryCreationAction,
    _original_movie_loan: MovieLoan,
) -> ExternResult<ValidateCallbackResult> {
    if action.author != original_action.author().clone() {
        return Ok(ValidateCallbackResult::Invalid("Agents can only delete their own MovieLoan records.".to_string()));
    }
    Ok(ValidateCallbackResult::Valid)
}
```

You can find other stub functions in that file for links that point to the most recent update (if you [chose that option](/get-started/3-forum-app-tutorial/#scaffold-most-recent-update-link) when you scaffolded the entry type), links that manage collections of entries, and backlinks from entries to entries that depend on them.

### `validate_agent_joining`

Use this function to validate the [**membrane proof**](/build/genesis-self-check-callback/#membrane-proof-a-joining-code-for-a-network). Note that this is different from `genesis_self_check`, in that it's called from the `validate` function so it can access DHT data.

> This example implements a simple invite code for a network that people can invite their friends to join. All that's required is the presence of an 'invite' action on the DHT, whose hash becomes the invite code. (It's not a very secure pattern; please don't duplicate this in high-security hApps.) As a bonus, it shows a good pattern where the code for basic pre-validation is shared with `genesis_self_check`.

Reviewer 3: ```suggestion
This example implements a simple invite code for a network that people can invite their friends to join. All that's required is the presence of an 'invite' action on the DHT, whose hash becomes the invite code. (It's not a very secure pattern; please don't duplicate this in high-security hApps.) Some of the business logic is shared with `genesis_self_check`.
```

> This example implements a simple invite code for a network that people can invite their friends to join. All that's required is the presence of an 'invite' action on the DHT, whose hash becomes the invite code. (It's not a very secure pattern; please don't duplicate this in high-security hApps.) As a bonus, it shows a good pattern where the code for basic pre-validation is shared with `genesis_self_check`.

Reviewer 3: ```suggestion
This example implements a simple invite code for a network that people can invite their friends to join. All that's required is the presence of an 'invite' action on the DHT, whose hash becomes the invite code. As a bonus, it shows a good pattern where the code for basic pre-validation is shared with `genesis_self_check`.
```

Why not just validate that the invitation includes the recipient agent? Would it be secure then, or is there something else missing?

Author: I wanted to avoid including the invitee's public key because then we'd have to get into talk about when a public key is generated vs when memproofs are requested. I decided to go with your suggestion but added a note about the race problem in https://github.com/holochain/docs-pages/pull/520/commits/5ca238acce48a98419adaa205d4f930a86123c60 .

Reviewer 3: > a note about the race problem in https://github.com/holochain/docs-pages/commit/5ca238acce48a98419adaa205d4f930a86123c60 .

This is a lot of additional info. I would rather leave that out here and talk about it when we get to app patterns.

Author: yes, this pattern does introduce a deep rabbit hole that I'd rather not go down either. Maybe I'll revert this like you suggest.

Author: Update: decided to keep the example -- it seems more readable somehow -- but removed the warning about deferred memproofs. https://github.com/holochain/docs-pages/pull/520/commits/138e890b7645bec86958c789d22fd131338aca60

```rust
use hdi::prelude::*;
use base64::prelude::*;

> pub fn validate_agent_joining(

Reviewer 3: It's definitely a bit confusing to me that this uses functions `validate_invite_code_format` and `decode_invite_code` that haven't been defined anywhere.

Also it doesn't look like any of the business logic has been shared with `genesis_self_check`?

Author: Both fixed in https://github.com/holochain/docs-pages/pull/520/commits/a9292c0d1954b6cf3850c5367aab62ba5bce09d2 . The absent functions was a bad paste, and I wanted to avoid including the invitee's public key because then we'd have to get into talk about when a public key is generated vs when memproofs are requested. I decided to go with your suggestion but added a note about the race problem in https://github.com/holochain/docs-pages/pull/520/commits/5ca238acce48a98419adaa205d4f930a86123c60 .

            let cert_is_valid = verify_signature(
                dna_props.authorized_joining_certificate_issuer,
                signature,
                data.agent_key
            )?;
            if cert_is_valid {
                return Ok(ValidateCallbackResult::Valid);
            } else {
                return Ok(ValidateCallbackResult::Invalid("Joining certificate wasn't valid. Please try entering it again or asking the certificate issuer for a new one.".into()));
            }
        }
    }
}
```

<!-- TODO: Write about inductive validation when it's not broken https://github.com/holochain/holochain/issues/4669 -->

## Reference

* [`holochain_integrity_types::op::Op`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/op/enum.Op.html)
* [`holochain_integrity_types::action::AgentValidationPkg`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/action/enum.Action.html#variant.AgentValidationPkg)
* [`holochain_integrity_types::genesis::GenesisSelfCheckData`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/genesis/type.GenesisSelfCheckData.html)

> * [`holochain_integrity_types::validate::ValidateCallbackResult`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/genesis/type.GenesisSelfCheckData.html)

Reviewer 3: ```suggestion
* [`holochain_integrity_types::validate::ValidateCallbackResult`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/validate/enum.ValidateCallbackResult.html)
```
